### ISSUE TITLE
ZFS built against system SPL rather than local SPL from repository

### ISSUE THREAD

**Issue by Harvie** [openzfs/zfs#8585]
BTW always when i am building the ZFS, it ends up building it against SPL currently running in system instead of against SPL from ./module/spl/ ... This is annoying and makes it very hard to build ZFS on system that already runs ZFS... Especialy when there is ZFS installed as part of kernel package, which cannot be uninstalled (eg. proxmox repository)

This leads to resulting ZFS module being unloadable with errors like this:

```
[   74.837630] icp: Unknown symbol __cv_wait (err -22)
[   74.852237] icp: disagrees about version of symbol __cv_broadcast
[   74.868063] icp: Unknown symbol __cv_broadcast (err -22)
[   74.883161] icp: disagrees about version of symbol __cv_signal
[   74.898734] icp: Unknown symbol __cv_signal (err -22)
[   74.913594] icp: disagrees about version of symbol __kstat_install
[   74.929522] icp: Unknown symbol __kstat_install (err -22)
```

Can you please modify ZFS makefile to build SPL from repo and then forcefuly link ZFS against it? instead of linking to some autodetected SPL that's present in system.

**Comment by c0d3z3r0** [comment]
@Harvie you should provide more information and use the template for any issues.... without it is very hard to reproduce this. I personally never had this problem.

**Comment by Harvie** [comment]
Maybe i am doing something wrong... I've done following on Debian:

  * Cloned zfs repo to /usr/src/zfs-git/
  * cd /usr/src/zfs-git/
  * ./autogen.sh
  * dkms add zfs/git
  * dkms autoinstall
  * depmod

I know this is not exactly reccomended way (as in https://github.com/zfsonlinux/zfs/wiki/Custom-Packages )
But i didn't liked the idea of using intermediate .deb package, when i was rebuilding this often.

It worked OK, but i had to remove all mentions of older ZFS version from system and rebuild, otherwise i got unloadable modules. Note that i've been doing this on Debian with Proxmox kernel, which has ZFS directly included in the kernel package, so there's no way of uninstalling distro version of ZFS module using package manager.

**Comment by c0d3z3r0** [comment]
You are missing `./configure --with-config=kernel` here, don't you?

**Comment by c0d3z3r0** [comment]
Hm. Just tested your workflow... how do you generate your dkms.conf?

**Comment by Harvie** [comment]
DKMS seems to call `./configure` automaticaly
I've forgot to mention `./scripts/dkms.mkconf -v git -f dkms.conf -n zfs`, i've edited the previous post to reflect that.

**Comment by c0d3z3r0** [comment]
Ah, yep, that looks fine...  **maybe** I have found the issue... trying to reproduce this here

**Comment by c0d3z3r0** [comment]
Could you run `ZFS_DKMS_ENABLE_DEBUG=y sudo -E dkms build zfs/git` without touching the systems zfs stuff and add the output here, please?

**Comment by Harvie** [comment]
I no longer have this problem, since i've manualy purged the old version. I will have to figure out how to reinstall distro zfs and reproduce again.

**Comment by Harvie** [comment]
Before:

```
# cat /lib/modules/4.15.18-12-pve/modules.dep | grep zfs
zfs/zpios.ko: zfs/zfs.ko zfs/zunicode.ko zfs/zavl.ko zfs/icp.ko zfs/zcommon.ko zfs/znvpair.ko zfs/spl.ko
zfs/zunicode.ko:
zfs/znvpair.ko: zfs/spl.ko
zfs/spl.ko:
zfs/zcommon.ko: zfs/znvpair.ko zfs/spl.ko
zfs/icp.ko: zfs/spl.ko
zfs/zavl.ko:
zfs/zfs.ko: zfs/zunicode.ko zfs/zavl.ko zfs/icp.ko zfs/zcommon.ko zfs/znvpair.ko zfs/spl.ko
zfs/splat.ko: zfs/spl.ko
```


After dkms install:

```
# cat /lib/modules/4.15.18-12-pve/modules.dep | grep zfs
zfs/zpios.ko: updates/dkms/zfs.ko updates/dkms/zunicode.ko updates/dkms/zlua.ko updates/dkms/zcommon.ko updates/dkms/znvpair.ko updates/dkms/zavl.ko zfs/icp.ko updates/dkms/spl.ko
zfs/icp.ko: updates/dkms/spl.ko
updates/dkms/zfs.ko: updates/dkms/zunicode.ko updates/dkms/zlua.ko updates/dkms/zcommon.ko updates/dkms/znvpair.ko updates/dkms/zavl.ko zfs/icp.ko updates/dkms/spl.ko
zfs/splat.ko: updates/dkms/spl.ko
```

This seems to me as mixing modules from `./zfs/` and `./updates/dkms/`

**Comment by Harvie** [comment]
Yes! I've rebooted and it failed again. So the logs attached in previous post are logs of ZFS being build with this bug. Gives me this error in dmesg:

```
[   87.030466] icp: disagrees about version of symbol __cv_destroy
[   87.043710] icp: Unknown symbol __cv_destroy (err -22)
[   87.056224] icp: disagrees about version of symbol __kstat_delete
[   87.069659] icp: Unknown symbol __kstat_delete (err -22)
[   87.082339] icp: disagrees about version of symbol __cv_init
[   87.095319] icp: Unknown symbol __cv_init (err -22)
[   87.107560] icp: disagrees about version of symbol __kstat_create
[   87.120983] icp: Unknown symbol __kstat_create (err -22)
[   87.133639] icp: disagrees about version of symbol __cv_wait
[   87.146619] icp: Unknown symbol __cv_wait (err -22)
[   87.158817] icp: disagrees about version of symbol __cv_broadcast
[   87.172226] icp: Unknown symbol __cv_broadcast (err -22)
[   87.184907] icp: disagrees about version of symbol __cv_signal
[   87.198061] icp: Unknown symbol __cv_signal (err -22)
[   87.210504] icp: disagrees about version of symbol __kstat_install
[   87.224018] icp: Unknown symbol __kstat_install (err -22)
[   87.626310] icp: disagrees about version of symbol __cv_destroy
[   87.639565] icp: Unknown symbol __cv_destroy (err -22)
[   87.652126] icp: disagrees about version of symbol __kstat_delete
[   87.665554] icp: Unknown symbol __kstat_delete (err -22)
[   87.678308] icp: disagrees about version of symbol __cv_init
[   87.691298] icp: Unknown symbol __cv_init (err -22)
[   87.703533] icp: disagrees about version of symbol __kstat_create
[   87.716943] icp: Unknown symbol __kstat_create (err -22)
[   87.729594] icp: disagrees about version of symbol __cv_wait
[   87.742548] icp: Unknown symbol __cv_wait (err -22)
[   87.754742] icp: disagrees about version of symbol __cv_broadcast
[   87.768149] icp: Unknown symbol __cv_broadcast (err -22)
[   87.780862] icp: disagrees about version of symbol __cv_signal
[   87.793998] icp: Unknown symbol __cv_signal (err -22)
[   87.806453] icp: disagrees about version of symbol __kstat_install
[   87.819970] icp: Unknown symbol __kstat_install (err -22)
```

**Comment by c0d3z3r0** [comment]
> Also dkms instal said:
> 
> ```
> icp.ko:
> Running module version sanity check.
> Error! Module version 0.8.0-rc3_141_gd93d4b1ac for icp.ko
> is not newer than what is already found in kernel 4.15.18-12-pve (0.8.0-rc3_141_gd93d4b1ac).
> You may override by specifying --force.
> ```
> This seems to me as mixing modules from `./zfs/` and `./updates/dkms/`

Yep. The problem is the same version. DKMS will not install modules with the same version AFAIK. You need to add a custom version addition by editing `META`. e.g. set "Release:       rc3-custom"

**Comment by Harvie** [comment]
But the problem occured even when there was no 0.8.0 installed previously...

**Comment by c0d3z3r0** [comment]
> But the problem occured even when there was no 0.8.0 installed previously...

Just try it ;) In the meanwhile I'm looking at your logs..

**Comment by Harvie** [comment]
I will just try to pull the latest git, it will change the version string...

**Comment by Harvie** [comment]
I did so and found out it happens even with new version. There's no way that this version was ever previously installed on the system:

```
icp.ko:
Running module version sanity check.
Error! Module version 0.8.0-rc3_148_gb31cf30a1 for icp.ko
is not newer than what is already found in kernel 4.15.18-12-pve (0.8.0-rc3_148_gb31cf30a1).
You may override by specifying --force.
```

**Comment by Harvie** [comment]
Any ideas what might cause this?

**Comment by Harvie** [comment]
bump

**Comment by behlendorf** [comment]
This issue may have been resolved by 4d6043f2b74ded58d36d3254acd12da16023c844 which is included in the 2.0.x releases.

**Comment by behlendorf** [comment]
Closing.  The spl and zfs repositories have been merged since the 0.8 releases removing this possible confusion at build time.

**Comment by telsasoft** [comment]
For those who find this issue after searching for the error message:

I had this issue after building+installing RPMs for zfs 2.2 on a centos7 system which was already running zfs 2.1 (from the zfsonlinux RPMs).

The solution was to rmmod the modules from zfs 2.1 (spl icp znvpair zcommon zavl zlua zunicode).
modprobe -r didn't work - not sure why.

It looks like the same issue here
https://answers.launchpad.net/ubuntu/+source/zfs-linux/+question/635917
